pages/en/developer/querying-from-your-app.mdx

```diff
@@ -2,135 +2,572 @@
+
+Looking at the following query, you will notice that some fields are repeated across multiple Selection-Sets (`{ ... }`):
+
+```graphql
+query {
+  bondEvents {
+    id 
+    newDelegate {
+      id
+      active
+      status
+    }
+    oldDelegate {
       id
-      tokenID
-      contentURI
-      metadataURI
+      active
+      status
+    }
+  }
+}
+```
+
+Such repeated fields (`id`, `active`, `status`) bring many issues:
+
+- harder to read for more extensive queries
+- when using tools that generate TypeScript types based on queries (*more on that in the last section*), `newDelegate` and `oldDelegate` will result in two distinct inline interfaces.
+
+A refactored version of the query would be the following:
+
+```graphql
+query {
+  bondEvents {
+    id 
+    newDelegate {
+      ...DelegateItem
+    }
+    oldDelegate {
+      ...DelegateItem
     }
   }
+}
+
+# we define a fragment (subtype) on Transcoder
+# to factorize repeated fields in the query
+fragment DelegateItem on Transcoder {
+  id
+  active
+  status
+}
+```
+
+Using GraphQL `fragment` will improve readability (especially at scale) but also will result in better TypeScript types generation.
+
+When using the types generation tool, the above query will generate a proper `DelegateItemFragment` type (*see last “Tools” section*).
+
+<br />
+
+### GraphQL Fragment do's and don'ts
+
+**Fragment base must be a type**
+
+A Fragment cannot be based on a non-applicable type, in short, **on type not having fields**:
+
+```graphql
+fragment MyFragment on BigInt {
+  # ...
+}
+```
+
+`BigInt` is a **scalar** (native “plain” type) that cannot be used as a fragment's base.
+
+**How to spread a Fragment**
+
+Fragments are defined on specific types and should be used accordingly in queries.
+
+Example:
+
+```graphql
+query {
+  bondEvents {
+    id 
+    newDelegate {
+      ...VoteItem # Error! `VoteItem` cannot be spread on `Transcoder` type
+    }
+    oldDelegate {
+      ...VoteItem
+    }
+  }
+}
+
+fragment VoteItem on Vote {
+  id
+  voter
+}
+```
+
+`newDelegate` and `oldDelegate` are of type `Transcoder`.
+
+It is not possible to spread a fragment of type `Vote` here.
+
+**Define Fragment as an atomic business unit of data**
+
+GraphQL Fragment must be defined based on their usage.
+
+For most use-case, defining one fragment per type (in the case of repeated fields usage or type generation) is sufficient.
+
+Here is a rule of thumb for using Fragment:
+
+- when fields of the same type are repeated in a query, group them in a Fragment
+- when similar but not the same fields are repeated, create multiple fragments, ex:
+
+```graphql
+# base fragment (mostly used in listing)
+fragment Voter on Vote {
+  id
+  voter
+}
+
+# extended fragment (when querying a detailed view of a vote)
+fragment VoteWithPoll on Vote {
+  id
+  voter
+  choiceID
+  poll {
+    id
+    proposal
+  }
+}
+```
+
+---
+
+## The essential tools
```

Comment:
Nice! I would add some browser based tools, like [Altair](https://altair.sirmuel.design/) or [GraphiQL](https://graphiql-online.com/graphiql) since they are a very low-overhead way to check queries for problems.

Reply:
Done in b656828